trusted_applet/internal/storage/slots/journal.go

```diff
@@ -0,0 +1,312 @@
+	return j.current.Data, j.current.Revision
+}
+
+// Update creates a new entry record in the journal for the provided data.
+// The new record's revision will be one greater than the previous record (or 1
+// if no previous record exists).
+func (j *Journal) Update(data []byte) error {
+	if l := len(data); l > int(j.maxDataBytes) {
+		return fmt.Errorf("attemping to write %d bytes, larger than the max permitted in this journal (%d bytes)", l, j.maxDataBytes)
+	}
+	h := sha256.Sum256(data)
+	e := entry{
+		Magic:      [4]byte{magic0[0], magic0[1], magic0[2], magic0[3]},
+		Revision:   j.current.Revision + 1,
+		DataLen:    uint64(len(data)),
+		DataSHA256: h,
+		Data:       data,
+	}
+
+	cap := (j.start + j.length - j.nextBlock) * j.dev.BlockSize()
+	if cap < uint(e.Size()) {
+		// The record won't fit in the remaining space, so wrap around and write at the beginning.
+		j.nextBlock = j.start
+	}
+	// TODO(al): consider making this more "streamy".
+	buf := &bytes.Buffer{}
+	if err := marshalEntry(e, buf); err != nil {
+		return fmt.Errorf("failed to marshal entry: %v", err)
+	}
+	if err := j.dev.WriteBlocks(j.nextBlock, buf.Bytes()); err != nil {
+		return fmt.Errorf("failed to write blocks: %v", err)
+	}
+	// Finally, we'll rescan the journal and make sure that all is well.
+	if err := j.init(); err != nil {
+		return fmt.Errorf("failed to re-scan journal: %v", err)
+	}
+	if got, want := j.current.Revision, e.Revision; got != want {
+		return fmt.Errorf("journal error, latest entry revision %d != written revision %d", got, want)
+	}
+	return nil
+}
+
+// Init scans the journal to figure out the latest valid record, if any.
+func (j *Journal) init() error {
+	// Start where all good stories do: at the beginning!
+	lba := j.start
+	var lastEntry entry
+	nextWriteLBA := j.start
+	for lba < j.start+j.length {
+		br := newBlockReader(j.dev, lba)
+		e, err := unmarshalEntry(br)
+		if err != nil {
+			if lastEntry.Revision > 0 {
+				glog.V(2).Infof("Scanned to invalid entry, using last good entry seen@rev %d (%v)", lastEntry.Revision, err)
+				// We already found the lastet record in the journal, so we're done.
+				break
+			}
+			// Since we haven't already found a good entry, and we were unable
+			// to unmarshal this one,
+			// this means that either:
+			//  a) the journal is completely empty, or
+			//  b) the previously good entry/ies at the start of the journal
+			//     have been completely or partially overwritten during a
+			//     failed write attempt.
+			// Either way, we don't have a valid entry wth a length field we can
+			// rely on, so we we'll have to fall back to scanning all blocks to
+			// look for one.
+			lba++
+			continue
+		}
+		if e.Revision > lastEntry.Revision {
+			glog.V(3).Infof("Scan found rev %d(@ block %d), continuing", e.Revision, lba)
+			// We've found a(nother) good entry, so update our state
+			lastEntry = *e
+			// Skip past the blocks we've just read
+			lba = (br.pos-1)/br.dev.BlockSize() + 1
```

Comment:
I fail to understand how lba can increase.

Let's take the values from TestWriteSizeLimit: start=1, blocksize = 32.
lba = (br.pos-1)/br.dev.BlockSize() + 1
from newBlockReader we have br.pos = j.start * dev.Blocksize(), i.r br.pos = 32
(32-1)/32+1 = 1 lba has not changed.

What am I missing on?

Reply:
Empirically, it definitely does increase :)

Ok, so: journal `start=1, bs=32`, that means initial `lba=1` and so `br.pos=32` before any reads, if we then read a full block `br.pos+=bs` == 64, so `lba=(64-1)/32+1` == `2`